gpii/node_modules/lifecycleManager/src/LifecycleManagerUtils.js

```diff
@@ -0,0 +1,106 @@
+/*!
+ * Lifecycle Manager
+ *
+ * Copyright 2019 OCAD University
+ *
+ * Licensed under the New BSD license. You may not use this file except in
+ * compliance with this License.
+ *
+ * The research leading to these results has received funding from the European Union's
+ * Seventh Framework Programme (FP7/2007-2013)
+ * under grant agreement no. 289016.
+ *
+ * You may obtain a copy of the License at
+ * https://github.com/GPII/universal/blob/master/LICENSE.txt
+ */
+
+"use strict";
+
+var fluid = fluid || require("infusion");
+var gpii = fluid.registerNamespace("gpii");
+
+(function () {
+    /*
+     * Remove settings that are not in the matchMaker settings results.
+     * @param {Array} settingValues - An array of objects with each object keyed by a solution id with its value being
+     * an object of setting values that this solution supports.
+     * @param {Object} matchMakerSolutionRegistryEntries - The solution registry entries in the calculated lifecyle
+     * instructions. Each solution registry entry contains a list of settings generated by a matchmaking process.
+     * @return {Object} An object with each element keyed by a solution id with its value only having settings
+     * generated by a matchmaking process.
+     */
+    gpii.lifecycleManager.getMatchMakerSettings = function (settingValues, matchMakerSolutionRegistryEntries) {
+        var settingsTogo = {};
+        // 1. collect all settings generated from the a matchmaking process
+        var matchedSettingsForSolutions = {};
+        fluid.each(matchMakerSolutionRegistryEntries, function (solutionSpec, solutionId) {
+            var matchedSettings = [];
+            fluid.each(solutionSpec.settingsHandlers, function (settingsHandlerSpec) {
+                matchedSettings = matchedSettings.concat(fluid.keys(settingsHandlerSpec.settings));
+            });
+            fluid.set(matchedSettingsForSolutions, [solutionId], matchedSettings);
+        });
+
+        // 2. remove setting values that are not in matchedSettings
+        fluid.each(settingValues, function (oneSolutionBlock) {
+            fluid.each(oneSolutionBlock, function (settingBlocks, solutionId) {
+                fluid.each(settingBlocks, function (oneSettingBlock) {
+                    fluid.each(oneSettingBlock.settings, function (value, key) {
+                        if (matchedSettingsForSolutions[solutionId].includes(key)) {
+                            fluid.set(settingsTogo, [solutionId, key], value);
+                        };
+                    });
+                });
+            });
+        });
+        return settingsTogo;
+    };
+
+    /*
+     * Inverse transform settings to common or application terms
```

Comment:
Inverse -> invert? We could do with a lot more comments here explaining the context in which case this is helpful and preferably some example payloads accepted by this function. We are usually trying to supply @typedef s rather than plain {Object} for JSdoc arguments these days but I appreciate this might be difficult for our rather freely-structured prefs documents - but do try to do this wherever it is feasible.